kythe/cxx/verifier/souffle_interpreter.cc

```diff
@@ -23,71 +23,123 @@
 #include "absl/strings/str_cat.h"
 #include "glog/logging.h"
 #include "interpreter/Engine.h"
+#include "kythe/cxx/verifier/assertions_to_souffle.h"
+#include "kythe/cxx/verifier/verifier.h"
 #include "souffle/RamTypes.h"
 #include "souffle/io/IOSystem.h"
 #include "third_party/souffle/parse_transform.h"
 
 namespace kythe::verifier {
 namespace {
-constexpr std::array<int, 4> kInputData = {1, 2, 2, 3};
-
 class KytheReadStream : public souffle::ReadStream {
  public:
   explicit KytheReadStream(
       const std::map<std::string, std::string>& rw_operation,
-      souffle::SymbolTable& symbol_table, souffle::RecordTable& record_table)
-      : souffle::ReadStream(rw_operation, symbol_table, record_table) {}
+      souffle::SymbolTable& symbol_table, souffle::RecordTable& record_table,
+      const AnchorMap& anchors, const Database& database)
+      : souffle::ReadStream(rw_operation, symbol_table, record_table),
+        anchors_(anchors),
+        database_(database) {
+    if (auto op = rw_operation.find("anchors"); op != rw_operation.end()) {
+      anchor_it_ = anchors_.cbegin();
+      database_it_ = database_.cend();
+    } else {
+      database_it_ = database_.cbegin();
+      anchor_it_ = anchors_.cend();
+    }
+    std::array<souffle::RamDomain, 5> fields = {0, 0, 0, 0, 0};
+    empty_vname_ = record_table.pack(fields.data(), fields.size());
+  }
 
  protected:
   souffle::Own<souffle::RamDomain[]> readNextTuple() override {
-    if (pos_ >= kInputData.size()) {
-      return nullptr;
+    if (database_it_ != database_.end()) {
+      auto* t = (*database_it_)->AsApp()->rhs()->AsTuple();
+      auto tuple = std::make_unique<souffle::RamDomain[]>(5);
+      souffle::RamDomain vname[5];
+      CopyVName(t->element(0)->AsApp()->rhs()->AsTuple(), vname);
+      tuple[0] = recordTable.pack(vname, 5);
+      tuple[1] = t->element(1)->AsIdentifier()->symbol();
+      if (auto* id = t->element(2)->AsIdentifier(); id != nullptr) {
+        tuple[2] = id->symbol();
+      } else {
+        CopyVName(t->element(2)->AsApp()->rhs()->AsTuple(), vname);
+        tuple[2] = recordTable.pack(vname, 5);
+      }
+      tuple[3] = t->element(3)->AsIdentifier()->symbol();
+      tuple[4] = t->element(4)->AsIdentifier()->symbol();
+      ++database_it_;
+      return tuple;
+    } else if (anchor_it_ != anchors_.end()) {
+      auto tuple = std::make_unique<souffle::RamDomain[]>(3);
+      tuple[0] = anchor_it_->first.first;
+      tuple[1] = anchor_it_->first.second;
+      souffle::RamDomain vname[5];
+      CopyVName(anchor_it_->second->AsApp()->rhs()->AsTuple(), vname);
+      tuple[2] = recordTable.pack(vname, 5);
+      ++anchor_it_;
+      return tuple;
     }
-    auto tuple = std::make_unique<souffle::RamDomain[]>(2);
-    tuple[0] = kInputData[pos_++];
-    tuple[1] = kInputData[pos_++];
-    return tuple;
   }
 
  private:
-  int pos_ = 0;
+  void CopyVName(kythe::verifier::Tuple* vname, souffle::RamDomain* target) {
```

Comment:
`souffle::RamDomain (&target)[5]` (as hideous as the syntax is, it's clear that this is a reference to an array of 5 elements not just a pointer).